File: ruby/fourrier/lib/fourrier/freqs.rb
module Fourrier
  module Private
    class Freqs
      def initialize(harmonics)
        @p = Fourrier::CInterface.allocFrequencies(harmonics)
      end
      
      def uninitialize
        
      end
      
      def ptr
        @p
      end
      
      def getHarmonics
        Fourrier::CInterface.getHarmonics(@p)
      end
      
      def getFrequency(h)
        Fourrier::CInterface.getFrequency(@p, h)
      end
      
      def getFrequencies
        freqs = []
        getHarmonics.times do |i|
          freqs << getFrequency(i)
        end
        freqs
      end
      
      def setFrequency(value, h)
        Fourrier::CInterface.setFrequency(@p, value, h)
      end
      
      def getAmplitude(h)
        Fourrier::CInterface.getAmplitude(@p, h)
      end
      
      def getAmplitudes
        amps = []
        getHarmonics.times do |i|
          amps << getAmplitude(i)
        end
        amps
      end
      
      def setAmplitude(value, h)
        Fourrier::CInterface.setAmplitude(@p, value, h)
      end
      
      def getPhase(h)
        Fourrier::CInterface.getPhase(@p, h)
      end
      
      def getPhases
        phases = []
        getHarmonics.times do |i|
          phases << getPhase(i)
        end
        phases
      end
      
      def setPhase(value, h)
        Fourrier::CInterface.setPhase(@p, value, h)
      end
      
    end # class Freqs
  end #module Private
end # module Fourrier

File: src/freqs.c
/**
* Fourrier
* A C library for tune computation
*
* Author: Cedric Hernaslteens <cedric.hernalsteens@cern.ch>
* 
* European Organization for Nuclear Research
*
* Copyright (c) 2014+ CERN. All rights reserved.
*
**/

#include <stdlib.h>
#include <stdbool.h>
#include <math.h>
#include <complex.h>

struct Freqs{
   double *frequencies;
   double *amplitudes;
   double *phases;
   double _Complex **basis;
   unsigned short int harmonics;
   unsigned int length;
   unsigned int basis_elements;
   _Bool allocated_basis;
};

struct Freqs* allocFrequencies(const unsigned short int h)
{
   struct Freqs* f = (struct Freqs*) malloc(sizeof(struct Freqs));
   f->harmonics = h;
   f->frequencies = (double*) malloc(h*sizeof(double));
   f->amplitudes = (double*) malloc(h*sizeof(double));
   f->phases = (double*) malloc(h*sizeof(double));
   return f;
}

short int allocFourierBasis(struct Freqs* f, const unsigned int length)
{
  if(!f->allocated_basis)
  {
    f->length = length;
    f->basis_elements = 0;
    f->basis = (double _Complex **) malloc(f->harmonics*sizeof(double*));
    int i = 0;
    while(i < f->harmonics)
    {
      f->basis[i] = (double _Complex*) malloc(length*sizeof(double));
    }
    f->allocated_basis = true;
  }
  return 1;
}

double _Complex*
addBasisElement(struct Freqs* f, const double freq)
{
  for(unsigned int i = 0; i < f->length; i++)
  {
    f->basis[f->basis_elements][i] = cexp(-I*i*freq*2*M_PI);
  }
  f->basis_elements++;
  return f->basis[f->basis_elements-1];
}

void freeFrequencies(struct Freqs* f)
{
   /* TO DO */
}

unsigned short int getHarmonics(const struct Freqs* f)
{
  return f->harmonics;
}

double getFrequency(const struct Freqs* f, const unsigned short int n)
{
   return f->frequencies[n];
}

void setFrequency(struct Freqs* f, const double freq, const unsigned short int n)
{
  f->frequencies[n] = freq;
}

double getAmplitude(const struct Freqs* f, const unsigned short int n)
{
   return f->amplitudes[n];
}

void setAmplitude(struct Freqs* f, const double amp, const unsigned short int n)
{
  f->amplitudes[n] = amp;
}

double getPhase(const struct Freqs* f, const unsigned short int n)
{
   return f->phases[n];
}

void setPhase(const struct Freqs* f, const double phase, const unsigned short int n)
{
  f->phases[n] = phase;
}

double _Complex* getVector(const struct Freqs* f, const unsigned short int n)
{
  return f->basis[n];
}